File: ios/Tradeguruelectrical/Services/JWTDecoder.swift
import Foundation

nonisolated enum JWTDecoder: Sendable {
    static func decode(_ jwt: String) -> (user: AuthUser, expiresAt: Date)? {
        let segments = jwt.split(separator: ".")
        guard segments.count == 3 else { return nil }

        var base64 = String(segments[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }

        guard let data = Data(base64Encoded: base64) else { return nil }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }

        guard let sub = json["sub"] as? String,
              let email = json["email"] as? String,
              let exp = json["exp"] as? TimeInterval else { return nil }

        let user = AuthUser(
            id: sub,
            email: email,
            name: json["name"] as? String,
            pictureURL: json["picture"] as? String
        )

        let expiresAt = Date(timeIntervalSince1970: exp)

        return (user: user, expiresAt: expiresAt)
    }
}

File: android/app/src/main/java/com/tradeguru/electrical/services/JWTDecoder.kt
package com.tradeguru.electrical.services

import android.util.Base64
import com.tradeguru.electrical.models.AuthUser
import org.json.JSONObject

object JWTDecoder {

    fun decode(jwt: String): AuthUser? {
        val segments = jwt.split(".")
        if (segments.size != 3) return null

        val payload = try {
            val padded = segments[1]
                .replace("-", "+")
                .replace("_", "/")
                .let { it + "=".repeat((4 - it.length % 4) % 4) }
            String(Base64.decode(padded, Base64.DEFAULT), Charsets.UTF_8)
        } catch (_: Exception) {
            return null
        }

        val json = try {
            JSONObject(payload)
        } catch (_: Exception) {
            return null
        }

        val sub = json.optString("sub").takeIf { it.isNotEmpty() } ?: return null
        val email = json.optString("email").takeIf { it.isNotEmpty() } ?: return null

        return AuthUser(
            id = sub,
            email = email,
            name = json.optString("name", null),
            pictureURL = json.optString("picture", null)
        )
    }

    fun getExpiry(jwt: String): Long? {
        val segments = jwt.split(".")
        if (segments.size != 3) return null

        val payload = try {
            val padded = segments[1]
                .replace("-", "+")
                .replace("_", "/")
                .let { it + "=".repeat((4 - it.length % 4) % 4) }
            String(Base64.decode(padded, Base64.DEFAULT), Charsets.UTF_8)
        } catch (_: Exception) {
            return null
        }

        val json = try {
            JSONObject(payload)
        } catch (_: Exception) {
            return null
        }

        val exp = json.optLong("exp", 0)
        if (exp <= 0) return null
        return exp * 1000
    }
}
